lib/ProductOpener/CRM.pm

```diff
@@ -0,0 +1,572 @@
+# Contact: contact@openfoodfacts.org
+# Address: 21 rue des Iles, 94100 Saint-Maur des Fossés, France
+#
+# Product Opener is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+=head1 NAME
+
+ProductOpener::CRM - manages integration with the Odoo CRM
+
+=head1 SYNOPSIS
+
+C<ProductOpener::CRM> contains functions to interact with the Odoo CRM
+
+
+=head1 DESCRIPTION
+
+For clarity:
+	- user/org refers to the pro platform side
+	- contact/company refers to the CRM side, where:
+			contact => Odoo partner is an 'individual'
+	  		company => Odoo partner is a 'company'
+
+=cut
+
+package ProductOpener::CRM;
+
+use ProductOpener::PerlStandards;
+use Exporter qw< import >;
+
+BEGIN {
+	use vars qw(@ISA @EXPORT_OK %EXPORT_TAGS);
+	@EXPORT_OK = qw(
+		&find_or_create_contact
+		&find_or_create_company
+		&add_contact_to_company
+		&create_opportunity
+		&add_user_to_company
+		&change_company_main_contact
+		&update_last_import_date
+		&update_last_export_date
+	);
+	%EXPORT_TAGS = (all => [@EXPORT_OK]);
+
+}
+use vars @EXPORT_OK;
+
+use ProductOpener::Config2;
+use ProductOpener::Tags qw/%country_codes_reverse/;
+use ProductOpener::Users qw/retrieve_user store_user/;
+use ProductOpener::Orgs qw/retrieve_org is_user_in_org_group/;
+use XML::RPC;
+use Log::Any qw($log);
+
+my %Odoo_tags = (Producer => '10',);
+
+our @api_credentials;
+our $xmlrpc;
+
+=head2 find_or_create_contact ($user_ref)
+
+Attempts to find a contact, and if it doesn't exist, creates it
+
+=head3 Arguments
+
+=head4 $user_ref 
+
+the user to which a 'contact' in the CRM should be linked
+
+=head3 Return values
+
+the id of the contact, or undef if an error occur
+
+=cut
+
+sub find_or_create_contact($user_ref) {
+	my $contact_id = find_contact($user_ref);
+	if (defined $contact_id) {
+		return if not link_user_with_contact($user_ref, $contact_id);
+	}
+	else {
+		$contact_id = create_contact($user_ref);
+	}
+	return $contact_id;
+}
+
+=head2 link_user_with_contact ($org_ref, $contact_id = undef)
+
+Set the off_username field of a contact to the user_id  
+
+=head3 Arguments
+
+=head4 $user_ref 
+
+=head4 $contact_id 
+
+=head3 Return values
+
+1 if success, undef otherwise
+
+=cut
+
+sub link_user_with_contact($user_ref, $contact_id) {
+	my $req = Odoo('res.partner', 'write',
+		[[$contact_id], {x_off_username => $user_ref->{userid}, category_id => [[4, $Odoo_tags{Producer}]]}]);
```

Comment:
What is 4? Maybe we should put it in a constant a the top of the module if it could be different on another Odoo instance